File: Medium/529 - Minesweeper/solution.ts
/**
 * Question 529. Minesweeper (https://leetcode-cn.com/problems/minesweeper/)
 */

const UNREVEALED_MINE: string = "M";
const UNREVEALED_EMPTY: string = "E";

const REVEALED_MINE: string = "X";
const REVEALED_BLANK: string = "B";

const UNREVEALED_SQUARES: string[] = [UNREVEALED_MINE, UNREVEALED_EMPTY];
const MINE_SQUARES: string[] = [REVEALED_MINE, UNREVEALED_MINE];

/**
 * This is the DFS solution, which is almost identical to the C# solution.
 *
 * 执行用时：156 ms, 在所有 typescript 提交中击败了 100.00% 的用户
 * 内存消耗：44.8 MB, 在所有 typescript 提交中击败了 100.00% 的用户
 *
 * @param {string[][]} board
 * @param {number[]} click
 * @returns {string[][]}
 */
function updateBoard(board: string[][], click: number[]): string[][] {
  const [clickedX, clickedY] = click;

  if (board[clickedX][clickedY] === UNREVEALED_MINE) {
    board[clickedX][clickedY] = REVEALED_MINE;
    return board;
  }

  let stack: number[][] = [click];

  while (stack.length) {
    const [currentX, currentY] = stack.pop() as number[];
    let adjacentMinesCount: number = 0;
    let adjacentUnrevealedSquares: number[][] = [];

    for (let xDirection = -1; xDirection <= 1; xDirection++) {
      for (let yDirection = -1; yDirection <= 1; yDirection++) {
        if (xDirection === yDirection && xDirection === 0) {
          continue;
        }

        const newX = currentX + xDirection;
        const newY = currentY + yDirection;

        if (
          newX >= 0 &&
          newX < board.length &&
          newY >= 0 &&
          newY < board[0].length &&
          UNREVEALED_SQUARES.some((square) => square === board[newX][newY])
        ) {
          adjacentUnrevealedSquares.push([newX, newY]);
          if (MINE_SQUARES.some((mine) => mine === board[newX][newY])) {
            adjacentMinesCount++;
          }
        }
      }
    }

    if (adjacentMinesCount === 0) {
      stack = [...stack, ...adjacentUnrevealedSquares];
      board[currentX][currentY] = REVEALED_BLANK;
    } else {
      board[currentX][currentY] = `${adjacentMinesCount}`;
    }
  }

  return board;
}

/**
 * 执行用时：140 ms, 在所有 typescript 提交中击败了 100.00% 的用户
 * 内存消耗：44.6 MB, 在所有 typescript 提交中击败了 100.00% 的用户
 *
 * @param {string[][]} board
 * @param {number[]} click
 * @returns {string[][]}
 */
function updateBoardBFS(board: string[][], click: number[]): string[][] {
  const [clickedX, clickedY] = click;

  if (board[clickedX][clickedY] === UNREVEALED_MINE) {
    board[clickedX][clickedY] = REVEALED_MINE;
    return board;
  }

  let queue: number[][] = [click];

  while (queue.length) {
    const [currentX, currentY] = queue.shift() as number[];
    let adjacentMinesCount: number = 0;
    let adjacentUnrevealedSquares: number[][] = [];

    for (let xDirection = -1; xDirection <= 1; xDirection++) {
      for (let yDirection = -1; yDirection <= 1; yDirection++) {
        if (xDirection === yDirection && xDirection === 0) {
          continue;
        }

        const newX = currentX + xDirection;
        const newY = currentY + yDirection;

        if (
          newX >= 0 &&
          newX < board.length &&
          newY >= 0 &&
          newY < board[0].length &&
          UNREVEALED_SQUARES.some((square) => square === board[newX][newY]) &&
          !queue.some(
            ([recordedX, recordedY]) => recordedX === newX && recordedY === newY
          )
        ) {
          adjacentUnrevealedSquares.push([newX, newY]);
          if (MINE_SQUARES.some((mine) => mine === board[newX][newY])) {
            adjacentMinesCount++;
          }
        }
      }
    }

    if (adjacentMinesCount === 0) {
      queue = [...queue, ...adjacentUnrevealedSquares];
      board[currentX][currentY] = REVEALED_BLANK;
    } else {
      board[currentX][currentY] = `${adjacentMinesCount}`;
    }
  }

  return board;
}

File: Medium/529 - Minesweeper/Solution.cs
/// Question 529. Minesweeper (https://leetcode-cn.com/problems/minesweeper/)

/// <summary>
/// 执行用时：456 ms, 在所有 C# 提交中击败了 34.78% 的用户
/// 内存消耗：35.4 MB, 在所有 C# 提交中击败了 46.15% 的用户
/// </summary>
public class Solution
{
    private const char UNREVEALED_MINE = 'M';
    private const char UNREVEALED_EMPTY = 'E';

    private const char REVEALED_MINE = 'X';
    private const char REVEALED_BLANK = 'B';

    private readonly List<char> UNREVEALED_SQUARES = new List<char> { UNREVEALED_MINE, UNREVEALED_EMPTY };
    private readonly List<char> MINE_SQUARES = new List<char> { REVEALED_MINE, UNREVEALED_MINE };

    public char[][] UpdateBoard(char[][] board, int[] click)
    {
        int x = click[0];
        int y = click[1];

        char clickedSquare = board[x][y];

        // Handle the situation if clicked square is an unrevealed mine.
        if (clickedSquare.Equals(UNREVEALED_MINE))
        {
            board[x][y] = REVEALED_MINE;
            return board;
        }

        Stack<(int, int)> stack = new Stack<(int, int)>();
        stack.Push((x, y));

        while (stack.Any())
        {
            (int currentX, int currentY) = stack.Pop();

            int adjacentMinesCount = 0;
            List<(int, int)> adjacentUnrevealedSquares = new List<(int, int)>();

            for (int xDirection = -1; xDirection <= 1; xDirection++)
            {
                for (int yDirection = -1; yDirection <= 1; yDirection++)
                {
                    if (xDirection == yDirection && xDirection == 0)
                    {
                        continue;
                    }

                    int newX = currentX + xDirection;
                    int newY = currentY + yDirection;

                    if (newX >= 0 && newX < board.Length &&
                        newY >= 0 && newY < board[0].Length &&
                        UNREVEALED_SQUARES.Contains(board[newX][newY]))
                    {
                        adjacentUnrevealedSquares.Add((newX, newY));

                        if (MINE_SQUARES.Contains(board[newX][newY]))
                        {
                            adjacentMinesCount++;
                        }
                    }
                }
            }


            if (!MINE_SQUARES.Contains(board[currentX][currentY]))
            {
                if (adjacentMinesCount == 0)
                {
                    board[currentX][currentY] = REVEALED_BLANK;
                    foreach (var adjacentUnrevealedSquare in adjacentUnrevealedSquares)
                    {
                        stack.Push(adjacentUnrevealedSquare);
                    }
                }
                else
                {
                    board[currentX][currentY] = $"{adjacentMinesCount}"[0];
                }
            }
        }

        return board;
    }
}